controllers/schedule_week.go

```diff
@@ -0,0 +1,409 @@
+package controllers
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/UniversityRadioYork/2016-site/models"
+	"github.com/UniversityRadioYork/2016-site/structs"
+	"github.com/UniversityRadioYork/2016-site/utils"
+	"github.com/UniversityRadioYork/myradio-go"
+	"github.com/gorilla/mux"
+)
+
+// weekFromVars extracts the year, and week strings from vars.
+func weekFromVars(vars map[string]string) (string, string, error) {
+	y, ok := vars["year"]
+	if !ok {
+		return "", "", errors.New("no year provided")
+	}
+	w, ok := vars["week"]
+	if !ok {
+		return "", "", errors.New("no week provided")
+	}
+
+	return y, w, nil
+}
+
+// weekdayFromVars extracts the year, week, and weekday strings from vars.
+func weekdayFromVars(vars map[string]string) (string, string, string, error) {
+	y, ok := vars["year"]
+	if !ok {
+		return "", "", "", errors.New("no year provided")
+	}
+	w, ok := vars["week"]
+	if !ok {
+		return "", "", "", errors.New("no week provided")
+	}
+	d, ok := vars["weekday"]
+	if !ok {
+		return "", "", "", errors.New("no weekday provided")
+	}
+
+	return y, w, d, nil
+}
+
+//
+// Week schedule algorithm
+// TODO(CaptainHayashi): move?
+//
+
+// WeekScheduleCell represents one cell in the week schedule.
+type WeekScheduleCell struct {
+	// Number of rows this cell spans.
+	// If 0, this is a continuation from a cell further up.
+	RowSpan uint
+
+	// Pointer to the timeslot in this cell, if any.
+	// Will be nil if 'RowSpan' is 0.
+	Item *structs.ScheduleItem
+}
+
+// WeekScheduleRow represents one row in the week schedule.
+type WeekScheduleRow struct {
+	// The hour of the row (0..23).
+	Hour int
+	// The minute of the show (0..59).
+	Minute int
+	// The cells inside this row.
+	Cells []WeekScheduleCell
+}
+
+// addCell adds a cell with rowspan s and item i to the row r.
+func (r *WeekScheduleRow) addCell(s uint, i *structs.ScheduleItem) {
+	r.Cells = append(r.Cells, WeekScheduleCell{RowSpan: s, Item: i})
+}
+
+// showStraddlesDay checks whether a show's start and finish cross over the boundary of a URY day.
+func showStraddlesDay(start, finish time.Time) bool {
+	nextDayStart := utils.StartOfDayOn(start.AddDate(0, 0, 1))
+	return finish.After(nextDayStart)
+}
+
+// calculateScheduleBoundaries works out the earliest and latest hours in the schedule that need to display.
+// It returns these as a pair of start and finish bound, both in terms of offsets from URY start time.
+func calculateScheduleBoundaries(items []structs.ScheduleItem) (sOffset, fOffset int, err error) {
+	if len(items) == 0 {
+		err = errors.New("calculateScheduleBoundaries: no schedule")
+		return
+	}
+
+	// These are the boundaries for culling, and are expanded upwards when we find shows that start earlier or finish later than the last-set boundary.
```

Comment:
This entire bit of code could be made a lot more efficient if we hold onto the first and last shows of each day.